Избирательное поведение

sigmarius opened the pull request:
- создана модель TasksSearch для поиска заданий по критериям из формы
- создана форма
- реализована фильтрация заданий в соответствии с выбранными в форме условиями

---
:mortar_board: [Избирательное поведение](https://up.htmlacademy.ru/yii/2/user/1247385/tasks/13)

sigmarius committed:
- создана модель TasksSearch для поиска заданий по критериям из формы
- создана форма
- реализована фильтрация заданий в соответствии с выбранными в форме условиями

VikingBO commented on fixtures/templates/tasks.php:
```diff
@@ -12,11 +12,11 @@
     'category_id' => $faker->numberBetween(1, 8),
     'client_id' => $faker->numberBetween(1, 5),
-    'worker_id' => $faker->numberBetween(1, 5),
+    'worker_id' => $faker->numberBetween(0, 5),
```
А 0 в каких случаях понадобится?

sigmarius replied:
0 это у меня в случаях, когда нет исполнителя. В БД у меня worker_id - INT NOT NULL сделан, и в миграции тоже.
Вот и перегенерировала данные, не подумала об этом с самого начала

VikingBO replied:
надо будет переделать, может к следующему заданию, миграцию поправить и сделать чуть посложнее логику в шаблоне, потому как идентификатор 0 тоже идентификатор, например в массивах же у нас первый числовой ключ `0`

VikingBO commented on models/TasksSearch.php:
```diff
@@ -0,0 +1,89 @@
+    /**
+     * Отключает правила валидации родительской модели
+     */
+    public function scenarios()
```
насколько я помню для отключения достаточно переопределить метод `rules()`, сценарии можно использовать для более детальной проработки модели, если брать те же `rules()` то для отдельных правил можно указывать конкретное название сценария к которому он будет применяться или наоборот к которому он не должен применяться.

sigmarius replied:
Это я взяла из примера в документации
![image](https://user-images.githubusercontent.com/57371471/216108493-f237f156-1b9e-4e4f-8354-57fb288f2e68.png)

VikingBO commented on models/TasksSearch.php:
```diff
@@ -0,0 +1,89 @@
+    public function calculateTimeDiff()
+    {
+        $now = new DateTime();
+        $now->modify("-$this->hoursPeriod hours");
```
очень плохая практика использовать двойные кавычки для  конкатенации данных, в частности для обработки в строке переменных, с версии 8.2 вроде такой способ уже не доступен к использованию.

sigmarius replied:
Исправила на обычную конкатенацию

VikingBO commented on models/TasksSearch.php:
```diff
@@ -0,0 +1,89 @@
+        // только поля определенные в rules() будут доступны для поиска
+        return [
+            ['categories', 'filter', 'filter' => function ($categories) {
+                return !empty($categories) ? array_map(fn ($category) => (int) $category, $categories) : '';
```
тут ты видимо вторым вариантом хотела возвращать пустой массив?

sigmarius replied:
Здесь у меня валилась ошибка, если категории не были выбраны по view, не помогло даже то, что я вроде как отключила передачу значений через 'unselect' => null. И у меня все заработало только с таким костыльком(
![image](https://user-images.githubusercontent.com/57371471/216110185-e6913bed-3c11-4b67-8fd7-6c41bc4000be.png)

VikingBO commented on models/TasksSearch.php:
```diff
@@ -0,0 +1,89 @@
+     */
+    public function scenarios()
+    {
+        return Model::scenarios();
```
метод `scenarios()` не является статическим, как подсказывает  IDE, поэтому он не может быть вызван статически.

sigmarius replied:
Тут я тоже ориентировалась на пример из документации
![image](https://user-images.githubusercontent.com/57371471/216111626-bfa54ccd-0170-4e4f-be36-816aaafdff85.png)

VikingBO commented on models/TasksSearch.php:
```diff
@@ -0,0 +1,89 @@
+        $query = Tasks::find()
+            ->where(['current_status' => TaskLogic::STATUS_NEW])
+            ->joinWith('city')
+            ->joinWith('category')
```
а вот эти joins мы зачем добавляем?

sigmarius replied:
Это я почитала в учебнике про ленивую и жадную загрузку связанных моделей - https://up.htmlacademy.ru/yii/2/module/5/item/9.
Посмотрела в дебагере, и поняла, что жадная лучше - с этими джойнами в дебагере выполняется 11 запросов к БД, а без них - 31

VikingBO commented on controllers/TasksController.php:
```diff
@@ -2,20 +2,23 @@
-            ->orderBy('published_at DESC')
-            ->all();
+        $categories = Categories::find()->asArray()->all();
+        $categories = ArrayHelper::map($categories, 'id', 'name');
```
а вот тут могу подсказать ещё один способ получения нужного массива:
```
Categories::find()->select('name')->indexBy('id')->asArray()->all();
```
Этот запрос тоже вернёт массив имён, и индексом этих имён будет идентификатор категории.

sigmarius replied:
Не получилось у меня, и никак не могу понять, почему?... Вроде бы пишу также - 
![image](https://user-images.githubusercontent.com/57371471/216116406-19d99070-f6a7-42bc-b4a1-3e6de5d9f16e.png)

А это результат от var_dump
![image](https://user-images.githubusercontent.com/57371471/216116582-1c2ae196-a9ed-4ce7-b9e8-32d42bf08199.png)

VikingBO replied:
Да, это я немного перепутал записи, не нужно ни asArray ни all, вот так будет работать   
```php
Categories::find()->select('name')->indexBy('id')->orderBy(['id' => SORT_ASC])->column()
```

VikingBO reviewed (changes requested):
Ну мне всё очень нравится, ты реально прям вникаешь в процесс, это здорово.
Посмотри вопросы или заметки что я оставил, ну и пиши что бы обсудить если надо.

sigmarius committed:
- исправляет ошибки

sigmarius commented:
Исправила что смогла

sigmarius committed:
- Делает поле worker_id в таблице Tasks не обязательным
- Добавляет миграцию на изменение таблицы в БД

sigmarius committed:
- Обновляет фикстуру для создания задач и заполнения таблицы Tasks в БД

sigmarius committed:
- Делает фикстуры из csv для таблиц Cities и Categories

sigmarius committed:
- По умолчанию значение worker_id не задано, условие добавляется к запросу при включенном чекбоксе на форме

sigmarius committed:
- Переносит константы из основного класса Task в собственные классы TaskStatuses и TaskActions
- Переименовывает директорию для задач
- Исправления с учетом workerId = null по дефолту

sigmarius committed:
- Исправляет фикстуру для Tasks
